### DISCUSSION TITLE
1st and 2nd derivative constraints on control inputs

### DISCUSSION THREAD

**Discussion by pavaninguva** [do-mpc/do-mpc#376] [Q&A] [answered]
Hi! 

What would be the best way to specify constraints on the first and second derivatives of the controller input, i.e.,

$$
lb \leq \frac{du}{dt} \leq ub \ \text{or} \left( \frac{du}{dt}  \right)^{2} \leq ub,
$$

$$
lb \leq \frac{d^{2}u}{dt^{2}} \leq ub  \ \text{or} \left( \frac{d^{2}u}{dt^{2}}  \right)^{2} \leq ub.
$$

The default formulation of the objective allows us to penalize sharp changes in $u$ using the `rterm` which in a sense captures the first derivative. For the second derivative, it seems like I would need to setup the nlp manually (based on #356 and #218). Would be grateful for any advice here. 

Also from a theoretical and practice POV, are there significant differences between implementing constraints on `_u` or penalizing sharp changes in `_u` using the `mterm`

**Comment by ccdonosoo** [comment, marked as answer]
Here is the way I do it. If you use the first derivative, you should give _u_k-1 as an state for the first derivative and _u_k-2 as another state for the second derivative. Then, you integrate with rhs, u_k-1_next = _u  and u_k-2 = u. After that, you could define d_uk as an algebraic equation and define the boundaries with the .bound method. For more details you could check my notebook.

https://colab.research.google.com/drive/1sXUsnHWsNonNQQTU49qHp2p7x-SyF3Iv?usp=sharing

**Reply by jrbull** [reply]
Is the penalty rterm applied to the scaled inputs? I have a lot of difficulty in getting the desired behaviour, my inputs are between O(1e5) and O(1e-3), and when I set what I think are appropriate rterm values - also quite wide-ranging - it causes numerical convergence problems.
